osu.Game.Rulesets.Mania/Edit/Blueprints/ManiaPlacementBlueprint.cs

```diff
@@ -0,0 +1,101 @@
+// Copyright (c) 2007-2018 ppy Pty Ltd <contact@ppy.sh>.
+// Licensed under the MIT Licence - https://raw.githubusercontent.com/ppy/osu/master/LICENCE
+
+using osu.Framework.Allocation;
+using osu.Framework.Graphics;
+using osu.Framework.Input.Events;
+using osu.Framework.Threading;
+using osu.Game.Rulesets.Edit;
+using osu.Game.Rulesets.Mania.Objects;
+using osu.Game.Rulesets.Mania.UI;
+using osu.Game.Rulesets.UI.Scrolling;
+using OpenTK;
+
+namespace osu.Game.Rulesets.Mania.Edit.Blueprints
+{
+    public class ManiaPlacementBlueprint<T> : PlacementBlueprint
+        where T : ManiaHitObject
+    {
+        protected new T HitObject => (T)base.HitObject;
+
+        /// <summary>
+        /// The current mouse position, snapped to the closest column.
+        /// </summary>
+        protected Vector2 SnappedMousePosition { get; private set; }
+
+        [Resolved]
+        private IManiaHitObjectComposer composer { get; set; }
+
+        [Resolved]
+        private IScrollingInfo scrollingInfo { get; set; }
+
+        public ManiaPlacementBlueprint(T hitObject)
+            : base(hitObject)
+        {
+            RelativeSizeAxes = Axes.None;
+        }
+
+        protected override bool OnMouseMove(MouseMoveEvent e)
+        {
+            updateSnappedPosition(e);
+            return true;
+        }
+
+        private ScheduledDelegate scheduledSnappedPositionUpdate;
+
+        private void updateSnappedPosition(MouseMoveEvent e)
+        {
+            scheduledSnappedPositionUpdate?.Cancel();
+            scheduledSnappedPositionUpdate = Schedule(() =>
+            {
+                Column column = ColumnAt(e.ScreenSpaceMousePosition);
+                if (column == null)
+                    SnappedMousePosition = e.MousePosition;
+                else
+                {
+                    // Snap to the column
+                    var parentPos = Parent.ToLocalSpace(column.ToScreenSpace(new Vector2(column.DrawWidth / 2, 0)));
+                    SnappedMousePosition = new Vector2(parentPos.X, e.MousePosition.Y);
+                }
+            });
+        }
+
+        protected double TimeAt(Vector2 screenSpacePosition)
```

Comment:
is this logic not able to be shared with selection blueprint (would be required there for movement, i think(

Reply:
You mean with `ManiaSelectionBlueprint.AdjustPosition()`?

You kind of lose purpose of that for TimeAt(), since each blueprint is still going to need to implement everything here minus only the return line. The blueprints differ in how they calculate the position - placement is relative to top-left of the hitobject container, whereas selection is relative to the anchor position of hitobjects.

Reply:
Furthermore the selection's method will have inter-column movement too